File: frontend/features/battle/mechanics/typeChart.ts
import type { BattleType } from "../data/types";
import { TYPES } from "../data/types";

export function getTypeMultiplier(
  attackType: BattleType,
  defenseType: BattleType
): number {
  if (attackType === defenseType) {
    return 1;
  }

  if (
    (attackType === TYPES.BLUE && defenseType === TYPES.RED) ||
    (attackType === TYPES.RED && defenseType === TYPES.GREEN) ||
    (attackType === TYPES.GREEN && defenseType === TYPES.BLUE) ||
    (attackType === TYPES.WHITE && defenseType === TYPES.BLACK) ||
    (attackType === TYPES.BLACK && defenseType === TYPES.WHITE)
  ) {
    return 2.0;
  }

  if (
    (attackType === TYPES.RED && defenseType === TYPES.BLUE) ||
    (attackType === TYPES.GREEN && defenseType === TYPES.RED) ||
    (attackType === TYPES.BLUE && defenseType === TYPES.GREEN)
  ) {
    return 0.5;
  }

  return 1;
}

File: backend/internal/battle/typechart.go
package battle

// GetTypeMultiplier はfeatures/battle/mechanics/typeChart.tsの移植。
// BLUE>RED>GREEN>BLUEの三すくみ、WHITE<->BLACKはお互いに2倍。YELLOWは今のところ相性なし。
func GetTypeMultiplier(attackType, defenseType BattleType) float64 {
	if attackType == defenseType {
		return 1
	}

	strong := (attackType == TypeBlue && defenseType == TypeRed) ||
		(attackType == TypeRed && defenseType == TypeGreen) ||
		(attackType == TypeGreen && defenseType == TypeBlue) ||
		(attackType == TypeWhite && defenseType == TypeBlack) ||
		(attackType == TypeBlack && defenseType == TypeWhite)

	if strong {
		return 2.0
	}

	weak := (attackType == TypeRed && defenseType == TypeBlue) ||
		(attackType == TypeGreen && defenseType == TypeRed) ||
		(attackType == TypeBlue && defenseType == TypeGreen)

	if weak {
		return 0.5
	}

	return 1
}
